### ISSUE TITLE
Forked tf script deadlocks unless disabling intra op parallelism

### ISSUE THREAD

**Issue by ceztko** [tensorflow/tensorflow#51832]
**System information**
- Have I written custom code: **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Ubuntu 20.04.2 LTS**
- Mobile device: **No**
- TensorFlow installed from: **pip**
- TensorFlow version (use command below): **v2.6.0-rc2-32-g919f693420e 2.6.0**
- Python version: **3.9.6**
- Bazel/GCC/Compiler version: **Not compiling from source**
- CUDA/cuDNN/GPU version: **Using CPU only**

**Describe the current behavior**

Me and the R&D team of our company are trying do a submission for an standardized benchmark, namely the [NIST FRVT 1:1](https://pages.nist.gov/frvt/html/frvt11.html) verification. Their benchmark suite employs this architecture:
- They run a custom submitter `initialize` function, where one can prepare the environment an run expensive functions that loads models and prepare temporary structures;
- They fork the process with unix `fork()` a number of times and they run either a `create_template` or `match` function in the forked children. According to the rules of the benchmark, the elaboration must be performed in **CPU** only. No GPU elaboration is allowed.

Since we can't modify the architecture because it's not in our control, we are trying to fit tensorflow so it will work according to these rules, but we are finding the children processes to deadlock trying to elaborate some layers. The only workaround we found is setting intra op parallelism to 1 with `tf.config.threading.set_intra_op_parallelism_threads(1)`, which appears to disable parallelism for operations like matrix multiplications. This workaround will not work in all scenarios, though. Running the model loading in the forked children will also workaround the issue but will penalize us in the benchmark since the time needed for the loading will be accounted for the elaboration.

**Describe the expected behavior**
Since we are enforcing CPU only elaboration and there are no resources that require exclusive access we are expecting tensorflow to correctly fit in this architecture and be able to run correctly in forked processes.

**Standalone code to reproduce the issue**
The following minimal python script will mimic the architecture and reproduce the issue. [Model](https://github.com/JonneOkkonen/MachineLearningProjects/blob/main/models/cats_vs_dogs_model_86_83.h5) and [test](https://github.com/JonneOkkonen/MachineLearningProjects/blob/main/testImages/cat.jpg) image used are linked.
It's available also as a Colab [notebook](https://colab.research.google.com/drive/1tvB9kZQvq6tiS7S4aMr761DQ9-DRcKkY?usp=sharing), which features a very similar behavior as running in a local machine, with the difference that in Colab notebook the `os.waitpid()` call never works but that could be an environment limitation.

```
import os
import tensorflow as tf
import cv2
import numpy

tf.config.set_visible_devices(tf.config.list_physical_devices('CPU'))
#tf.config.threading.set_intra_op_parallelism_threads(1) # Decomment this line and the child will not deadlock
#tf.debugging.set_log_device_placement(True)             # Decomment to see job placements

model = None

def initialize():
    global model
    model = tf.keras.models.load_model('cats_vs_dogs_model_86_83.h5')
    print('Model Loaded')

def child():
    print('Child spawned')
    imageSize = 128
    testImage = cv2.resize(src=cv2.imread('cat.jpg'), dsize=(imageSize, imageSize), interpolation=cv2.INTER_LINEAR) / 255
    result = model(testImage.reshape(-1, imageSize, imageSize, 3))[0]
    print('Result: Cat: ' + str(result[0]) + '| Dog: ' + str(result[1]))
    print('Child finished')

def parent():
    initialize()
    newpid = os.fork()
    if newpid == 0:
        child()
    else:
        pids = (os.getpid(), newpid)
        print("Parent: %d, Waiting for child: %d\n" % pids)
        os.waitpid(newpid, 0)

parent()
```
**Other info / logs**
Enabling log device placement will reveal that the elaboration stops elaborating in a convolution layer but with other scenarios the elaboration can deadlock in different operations.

```
Child spawned
2021-09-04 12:10:03.207086: I tensorflow/core/common_runtime/eager/execute.cc:1161] Executing op _EagerConst in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-04 12:10:03.209115: I tensorflow/core/common_runtime/eager/execute.cc:1161] Executing op Cast in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-04 12:10:03.210155: I tensorflow/core/common_runtime/eager/execute.cc:1161] Executing op ReadVariableOp in device /job:localhost/replica:0/task:0/device:CPU:0
2021-09-04 12:10:03.211392: I tensorflow/core/common_runtime/eager/execute.cc:1161] Executing op Conv2D in device /job:localhost/replica:0/task:0/device:CPU:0
```

We are using Numpy version: 1.19.5 as installed automatically from pip. We also tried to enforce updated numpy 1.21.2 as suggested [here](https://github.com/tensorflow/tensorflow/issues/13802#issuecomment-881665082) in an other `fork()` related issue but that didn't help.

**[Contributing](https://www.tensorflow.org/community/contribute)**
- Do you want to contribute a PR? (yes/no): **No**

**Comment by ceztko** [comment]
I was able to create a dump of the stuck process with gcore and I inspected it with gdb. From the analysis it shows a single thread stuck on a condition variable doing an Eigen operation. It follows an excerpt, attached the full [backtrace](https://github.com/tensorflow/tensorflow/files/7111317/all-threads-stacks.txt):

```
(gdb) thread apply all bt

Thread 1 (Thread 0x7fa1751d7740 (LWP 11903)):
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7ffea44694f8) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x7ffea44694a8, cond=0x7ffea44694d0) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x7ffea44694d0, mutex=0x7ffea44694a8) at pthread_cond_wait.c:638
#3  0x00007fa133c8be30 in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007fa1249054b3 in Eigen::Barrier::Wait() () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#5  0x00007fa1279d6700 in void Eigen::TensorEvaluator<Eigen::TensorContractionOp<Eigen::array<Eigen::IndexPair<long>, 1ul> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const, Eigen::NoOpOutputKernel const> const, Eigen::ThreadPoolDevice>::evalProductImpl<Eigen::TensorEvaluator<Eigen::TensorContractionOp<Eigen::array<Eigen::IndexPair<long>, 1ul> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const, Eigen::NoOpOutputKernel const> const, Eigen::ThreadPoolDevice>::NoCallback, 0>(float*, Eigen::TensorEvaluator<Eigen::TensorContractionOp<Eigen::array<Eigen::IndexPair<long>, 1ul> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const, Eigen::NoOpOutputKernel const> const, Eigen::ThreadPoolDevice>::NoCallback) const () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#6  0x00007fa1279d8caf in Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<long, 4> const, Eigen::TensorContractionOp<Eigen::array<Eigen::IndexPair<long>, 1ul> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const, Eigen::NoOpOutputKernel const> const> const> const, Eigen::ThreadPoolDevice, true, (Eigen::internal::TiledEvaluation)0>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<long, 4> const, Eigen::TensorContractionOp<Eigen::array<Eigen::IndexPair<long>, 1ul> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorImagePatchOp<-1l, -1l, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const> const, Eigen::TensorReshapingOp<Eigen::DSizes<long, 2> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer> const> const, Eigen::NoOpOutputKernel const> const> const> const&, Eigen::ThreadPoolDevice const&) () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#7  0x00007fa1279d96fc in void tensorflow::functor::SpatialConvolutionFunc<Eigen::ThreadPoolDevice, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::NoOpOutputKernel>(Eigen::ThreadPoolDevice const&, Eigen::TensorMap<Eigen::Tensor<float, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, long>, 16, Eigen::MakePointer>, int, int, int, int, Eigen::PaddingType const&, Eigen::NoOpOutputKernel const&, int, int, int, int) () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#8  0x00007fa127af62e1 in tensorflow::(anonymous namespace)::LaunchGeneric<Eigen::ThreadPoolDevice, float>::operator()(tensorflow::OpKernelContext*, tensorflow::Tensor const&, tensorflow::Tensor const&, int, int, int, int, tensorflow::Padding const&, std::vector<long, std::allocator<long> > const&, tensorflow::Tensor*, tensorflow::TensorFormat) [clone .isra.8556] [clone .constprop.8631] () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#9  0x00007fa127af6a11 in tensorflow::LaunchConv2DOp<Eigen::ThreadPoolDevice, float>::operator()(tensorflow::OpKernelContext*, bool, bool, tensorflow::Tensor const&, tensorflow::Tensor const&, int, int, int, int, tensorflow::Padding const&, std::vector<long, std::allocator<long> > const&, tensorflow::Tensor*, tensorflow::TensorFormat) () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
#10 0x00007fa127af786c in tensorflow::Conv2DOp<Eigen::ThreadPoolDevice, float>::Compute(tensorflow::OpKernelContext*) () from /home/ceztko/projects/current/ENVision/build-enpython/enpython/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so
[...] 
```

The presence of a singe thread probably means that a resource, possibly a thread pool, that is supposed to exist actually doesn't in the forked children, which is consistent with the fact that unix `fork()` will preserve only a single thread in the children. Another discovery is that by just moving the loading of the model to the child, but keeping all the imports as before as done [here](https://colab.research.google.com/drive/1tvB9kZQvq6tiS7S4aMr761DQ9-DRcKkY#scrollTo=SuUkj5vtTbiJ), will also fix the issue but is not desirable for complex models which takes long to load.

Because of this two considerations it appears the problem is either that tensorflow/Eigen can't recreate/reset the needed resources, possibly threads/condition variables, in the forked children (operations like these are usually done by fork handlers with `pthread_atfork`). Or alternatively just loading the model shouldn't cause tensorflow to load threading support in the library, which appear to happen here (I put breakpoints in `tensorflow.python.eager.context.ensure_initialized`).

**Comment by sushreebarsa** [comment]
@ceztko Could you please have a look at the similar [issue1](https://github.com/huggingface/transformers/issues/5486), [issue2](https://stackoverflow.com/questions/41233635/meaning-of-inter-op-parallelism-threads-and-intra-op-parallelism-threads) and let us know if it helps?Thanks!

**Comment by ceztko** [comment]
@sushreebarsa thank your for the prompt answer. [Issue1](https://github.com/huggingface/transformers/issues/5486) does not appear to matter here (I tried the TOKENIZERS_PARALLELISM=false, though, and it didn't help). Instead [issue2](https://stackoverflow.com/questions/41233635/meaning-of-inter-op-parallelism-threads-and-intra-op-parallelism-threads) is about the correct use of `set_intra_op_parallelism_threads()` and `set_inter_op_parallelism_threads()`, which I already tried both. The only one that seems to partially workaround the issue is `set_intra_op_parallelism_threads(1)`, which by definition should not enable threading for matrix operations, but that doesn't work always, for example evaluating the tensor with `model.predict()` also results in deadlock in my scenario with both `set_intra_op_parallelism_threads(1)`, `set_inter_op_parallelism_threads(1)` set. Before creating the bug report I searched several [similar](https://github.com/tensorflow/tensorflow/issues/13802) [issues](https://github.com/keras-team/keras/issues/9964) and solutions, but they mostly not apply here since the `fork()` call in my use case is performed outside of tensorflow itself and I can't change or override that call. My latest debugging [here](https://github.com/tensorflow/tensorflow/issues/51832#issuecomment-913122756) shows that the issue seems to be really related to the way tensorflow or Eigen handle (or not handle) the fork call.

**Comment by sushreebarsa** [comment]
@sanatmpa1 Was able to reproduce the issue on Colab using TF v2.5, 2.6 and tf-nightly,Please find the[ gist](https://colab.research.google.com/gist/sushreebarsa/dad6e56c8fa248ebdc10dccbad66b893/untitled190.ipynb#scrollTo=LUTrz0LeJAXK) for reference .Thanks!

**Comment by cantonios** [comment]
Eigen and TF's threadpool will not play nicely with unix `fork()`.  Calling fork while multiple threads are using a `std::mutex` will lead to undefined behavior.  We re-use the intra-op threadpool, so once the context is initialized, the threadpool will continue to exist.  It looks like calling `fork()` after this initialization is causing this issue.

You can try re-initializing the context in the child process, but that will likely introduce new issues.

Is there no way to ignore the model load time during benchmarking?

**Comment by ceztko** [comment]
I understand that, in the general, `fork()` can cause big headaches for all software using thread pools, and not all scenarios can be possibly supported because of unpredictable results if a fork() arises during some thread computations. Please note the following considerations:
- In this scenario we are forking after a model loading, so it should be a safe "still balls" moment;
- we really **can't** move the model loading in the child. It's very slow and penalize us in the benchmark. Even if move the model loading in a fork handler as a very bad trick, and this is not taken into account in the timings of the prediction operation, that would create harm in the execution of the benchmark anyway, being heavily slowed down while processing thousands and thousands of images;
- we evaluated alternative stacks and and we found that pytorch supports the load + fork execution model. In fact we found fork handlers being set in pythorch source code resetting their thread pool.

If you currently don't plan to fully support this scenario with threading enabled by implementing similar fork [handlers](https://man7.org/linux/man-pages/man3/pthread_atfork.3.html) we at least expect tensorflow to play nice when threading is fully disabled. This does not happen all the times. In [this](https://colab.research.google.com/drive/1tvB9kZQvq6tiS7S4aMr761DQ9-DRcKkY#scrollTo=iV9MTC0Rd2cI&line=29&uniqifier=1) script we both set intra/inter operations thread count to 1 and we perform the prediction with `model.predict()` instead. The script still hangs during and inference operation:
```
Executing op __inference_predict_function_527 in device /job:localhost/replica:0/task:0/device:CPU:0
```

Having a reliable single thread execution of tensorflow with this load+fork execution model would be enough for us since this benchmark is spawned in multiple processes anyway  and the benchmark also recommends it. Please let us know if you also reproduce the issue.

**Comment by cantonios** [comment]
Model loading initializes the context, which initializes the threadpool(s).  I don't see an easy way to separate this atm.

We will likely never implement fork handlers - that is specific to `pthread`.

We don't have a way to fully disable threading entirely - we have at least one thread in each pool (inter/intra).  This is necessary since we have asynchronous ops.

The only work-around I can think of is to provide a mechanism to re-initialize the threadpools that you can explicitly call at the start of your child processes.

We would be happy to review a pull-request along those lines.  Otherwise, this is a very specific unsupported use-case, so I can't make any guarantees about timelines.

**Comment by ceztko** [comment]
> Model loading initializes the context, which initializes the threadpool(s). I don't see an easy way to separate this atm.

Yes, we noticed it. It happens all the times model loading is performing some evaluation.

> We will likely never implement fork handlers - that is specific to `pthread`.

I don't understand: in all the platforms where `fork()` is available and functional (basically server/desktop unices, including MacOS) we expect `pthread_atfork` to be [functional](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/pthread_atfork.3.html) as [well](https://www.freebsd.org/cgi/man.cgi?query=pthread_atfork&sektion=3&manpath=FreeBSD+6.0-RELEASE), so why fork handlers can't be set up just on these platforms? Also pthread dependency is very likely put in indirectly by [other](https://github.com/tensorflow/tensorflow/issues/13802#issuecomment-881665082) depedendencies.

> We don't have a way to fully disable threading entirely - we have at least one thread in each pool (inter/intra). This is necessary since we have asynchronous ops.
> 

I understand, even if this sounds more like a design choice (either of Eigen or TF) than a true requirement.

> The only work-around I can think of is to provide a mechanism to re-initialize the threadpools that you can explicitly call at the start of your child processes. [...]
> We would be happy to review a pull-request along those lines.

We unfortunately don't have resources to put on this atm so our choice is just dodge the issue or continue the evaluation of other stacks where this scenario is better supported.

> Otherwise, this is a very specific unsupported use-case, so I can't make any guarantees about timelines.

While it may definitely be a specific use-case, we spotted several [users](https://stackoverflow.com/questions/54820729/sharing-saved-model-of-tensorflow-among-child-processes#comment122014575_54820729) that were definitely [attempting](https://stackoverflow.com/questions/52310182/how-to-handle-a-tensorflow-session-in-a-multiprocessing-environment-with-fork) to use tensorflow in this exact way. Also, because python has several limitations with threading, properly supporting this scenario may be a quick and elegant solution to create a pool of workers in all the platforms where `fork()` is available and cheap to call, so we recommend TF devs to reconsider their supporting intentions with regard to this use-case.

**Comment by cantonios** [comment]
Revisiting this briefly...

How many forks will there be?  I don't think we want to re-initialize the pools with the same parameters after fork, since that will increase the total number of threads and lead to over-subscription (and depending on the number of forks, can do so drastically).  We therefore essentially need to disable the threadpools altogether - but at that point, what is the purpose of the benchmark?

**Comment by ceztko** [comment]
Sorry for the late answer. Since then we moved to the "other" concurrent framework, which handles this scenario better, but I will do my best still to answer your questions and influence positively the development of tensorflow.

> but at that point, what is the purpose of the benchmark?

The [FRVT](https://pages.nist.gov/frvt/html/frvt11.html) benchmark is aiming at the score performance, not the speed. They time the executions but it's not much important as the score. They recommend (but they don't actually enforce it) to elaborate single thread since they will use a CPU only fork model to distribute the computation. It's a reasonable requirement to me and possibly scales much better in their environment.

> How many forks will there be?

This is not something we can control.

> I don't think we want to re-initialize the pools with the same parameters after fork, since that will increase the total number of threads

I think your concern is licit.

> We therefore essentially need to disable the threadpools altogether

If I understand it correctly this is something I already recommended in [this](https://github.com/tensorflow/tensorflow/issues/51832#issuecomment-926176681) reply. Having `set_intra_op_parallelism_threads(1)`, `set_inter_op_parallelism_threads(1)` not creating any threadpool (if it wasn't already initialized) would be great and accomodate the fork model greatly. Also after the fork you may have the option to create the threadpool at that point, for example with a small number of threads, to benefit both from multiprocessing and threading.

**Comment by cantonios** [comment]
@ceztko, sorry for the late response.

I may have a work-around if you're still interested.  The basic idea is (a) disable use of a global threadpool using an environment variable and (b) use an internal API to reset the context:
```python
# Import internal context API.
from tensorflow.python.eager import context

# Disable global threadpool re-use.
os.environ["TF_OVERRIDE_GLOBAL_THREADPOOL"] = "1"
#...

def child():
  # Reset context.
  context._reset_context()
  tf.config.threading.set_inter_op_parallelism_threads(1)
  tf.config.threading.set_intra_op_parallelism_threads(1)
  # ...
```

This should cause *most* of the threadpools to re-initialize in the child.  It works with your example [[colab](https://colab.research.google.com/drive/1odBaG7iXUMSb8Xlydc0kOkUjyXDh7_P9?usp=sharing)], and hopefully it will work in a real use-case as well.

**Comment by SuryanarayanaY** [comment]
Hi @ceztko ,

I am eager to know whether the proposed workaround in above [comment](https://github.com/tensorflow/tensorflow/issues/51832#issuecomment-1158431103) works for your case.

I do also validated the example code with TF2.11 version and its working fine as per attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/c9c496759c582b9338fd51ab60b6dff9/51832_tensorflowfork.ipynb). Please cross check and comeback if this is still an issue. Thanks!

**Comment by chloe-alverti** [comment]
Hello to all, while the workaround seemed to work up until python3.11, in python3.12 it is again broken. Any idea why and how it could be solved?

**Comment by alvinwong64** [comment]
@cantonios I'm also trying to submit for the NIST FRVT 1:1 benchmark
my method is embedding my python module in C++ with the use of pybind11

Where i try to mimic the action of 

IN C++ i do 

initialize{
        setenv("TF_OVERRIDE_GLOBAL_THREADPOOL", "1", 1); 
        m_tfModule = py::module_::import("tensorflow");
        m_tfModule.attr("config").attr("threading").attr("set_intra_op_parallelism_threads")(1);
        m_tfModule.attr("config").attr("threading").attr("set_inter_op_parallelism_threads")(1);
        m_cropModule = py::module_::import("crop");
        m_cropModel = m_tfModule.attr("keras").attr("models").attr("load_model")(cropModelPath,**kwargs);

}

and while at the 
createFaceTemplete{
        cropResult = m_cropModule.attr("process")(imageDict, m_cropModel);
}

So in the python module, this is what crop.process do:

process{
    context._reset_context()
    tf.config.threading.set_intra_op_parallelism_threads(1)
    tf.config.threading.set_inter_op_parallelism_threads(1)
    img = convert_raw_image(image_dict)
    img_raw = img.copy()
    img, pad_params = preprocessing(img, max_steps=max([8, 16, 32]))
    outputs = model(img[np.newaxis, ...]).numpy()
    outputs = postprocessing(outputs, pad_params)
}

Basically the flow of the whole program is

In C++, we call the initialize first where we will load our models, crop module, and etc...
Then will do the fork and run the createFaceTemplete where from the createFaceTemplete i call the python crop.process()

Now the question is :
While executing to obtain the model predictions, exception is thrown

Exception in crop module for image index 0: TypeError: <tf.Tensor 'inputs:0' shape=(1, 608, 480, 3) dtype=float32> is out of scope and cannot be used here. Use return values, explicit Python locals or TensorFlow collections to access it.

From what i understand is during the process , i do the context._reset_context() where it will clears the tensor threads? I'm kinda lost and not sure what to do to make my approach work.
